src/System.Memory/src/System/Text/EncodingExtensions.cs

```diff
@@ -0,0 +1,655 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Text
+{
+    public static class EncodingExtensions
+    {
+        /// <summary>
+        /// The maximum number of input elements after which we'll begin to chunk the input.
+        /// </summary>
+        /// <remarks>
+        /// The reason for this chunking is that the existing Encoding / Encoder / Decoder APIs
+        /// like GetByteCount / GetCharCount will throw if an integer overflow occurs. Since
+        /// we may be working with large inputs in these extension methods, we don't want to
+        /// risk running into this issue. While it's technically possible even for 1 million
+        /// input elements to result in an overflow condition, such a scenario is unrealistic,
+        /// so we won't worry about it.
+        /// </remarks>
+        private const int MAX_INPUT_ELEMENTS_PER_ITERATION = 1 * 1024 * 1024;
+
+        /// <summary>
+        /// Encodes the specified <see cref="ReadOnlySpan{Char}"/> to <see langword="byte"/>s using the specified <see cref="Encoding"/>
+        /// and writes the result to <paramref name="writer"/>.
+        /// </summary>
+        /// <param name="encoding">The <see cref="Encoding"/> which represents how the data in <paramref name="chars"/> should be encoded.</param>
+        /// <param name="chars">The <see cref="ReadOnlySequence{Char}"/> to encode to <see langword="byte"/>s.</param>
+        /// <param name="writer">The buffer to which the encoded bytes will be written.</param>
+        /// <exception cref="EncoderFallbackException">Thrown if <paramref name="chars"/> contains data that cannot be encoded and <paramref name="encoding"/> is configured
+        /// to throw an exception when such data is seen.</exception>
+        public static long GetBytes(this Encoding encoding, ReadOnlySpan<char> chars, IBufferWriter<byte> writer)
+        {
+            if (encoding is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.encoding);
+            }
+
+            if (writer is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.writer);
+            }
+
+            if (chars.Length <= MAX_INPUT_ELEMENTS_PER_ITERATION)
+            {
+                // The input span is small enough where we can one-shot this.
+
+                int byteCount = encoding!.GetByteCount(chars);
+                Span<byte> scratchBuffer = writer!.GetSpan(byteCount);
+                Debug.Assert(scratchBuffer.Length >= byteCount, "Scratch buffer was shorter than expected.");
```

Comment:
This doesn't look like a valid assert, given that `writer` is an arbitrary implementation that's passed in.  If we need to validate this, we should check for it and throw if it's wrong.